pkg/sources/filesystem/filesystem.go

```diff
@@ -83,100 +82,122 @@ func (s *Source) WithFilter(filter *common.Filter) {
 // Chunks emits chunks of bytes over a channel.
 func (s *Source) Chunks(ctx context.Context, chunksChan chan *sources.Chunk) error {
 	for i, path := range s.paths {
+		logger := ctx.Logger().WithValues("path", path)
+		if common.IsDone(ctx) {
+			return nil
+		}
 		s.SetProgressComplete(i, len(s.paths), fmt.Sprintf("Path: %s", path), "")
 
 		cleanPath := filepath.Clean(path)
-		done := false
-		go func() {
-			<-ctx.Done()
-			done = true
-		}()
-
-		err := fs.WalkDir(os.DirFS(cleanPath), ".", func(relativePath string, d fs.DirEntry, err error) error {
-			if err != nil {
-				return nil
-			}
-
-			path := filepath.Join(cleanPath, relativePath)
-			logger := ctx.Logger().WithValues("file", path)
-
-			fileStat, err := os.Stat(path)
-			if err != nil {
-				logger.Error(err, "unable to stat file")
-				return nil
-			}
-			if !fileStat.Mode().IsRegular() {
-				return nil
-			}
-
-			if s.filter != nil && !s.filter.Pass(path) {
-				return nil
-			}
-
-			inputFile, err := os.Open(path)
-			if err != nil {
-				logger.V(2).Info("error opening file", "error", err)
-				return nil
-			}
-			defer inputFile.Close()
-			logger.V(5).Info("scanning file")
-
-			reReader, err := diskbufferreader.New(inputFile)
-			if err != nil {
-				logger.Error(err, "Could not create re-readable reader.")
-			}
-			defer reReader.Close()
-
-			chunkSkel := &sources.Chunk{
-				SourceType: s.Type(),
-				SourceName: s.name,
-				SourceID:   s.SourceID(),
-				SourceMetadata: &source_metadatapb.MetaData{
-					Data: &source_metadatapb.MetaData_Filesystem{
-						Filesystem: &source_metadatapb.Filesystem{
-							File: sanitizer.UTF8(path),
-						},
-					},
-				},
-				Verify: s.verify,
-			}
-			if handlers.HandleFile(ctx, reReader, chunkSkel, chunksChan) {
-				return nil
-			}
-
-			if err := reReader.Reset(); err != nil {
-				return err
-			}
-			reReader.Stop()
-			data, err := io.ReadAll(reReader)
-			if err != nil {
-				return err
-			}
-			chunksChan <- &sources.Chunk{
-				SourceType: s.Type(),
-				SourceName: s.name,
-				SourceID:   s.SourceID(),
-				Data:       data,
-				SourceMetadata: &source_metadatapb.MetaData{
-					Data: &source_metadatapb.MetaData_Filesystem{
-						Filesystem: &source_metadatapb.Filesystem{
-							File: sanitizer.UTF8(path),
-						},
-					},
-				},
-				Verify: s.verify,
-			}
-			return nil
-		})
+		fileInfo, err := os.Stat(cleanPath)
+		if err != nil {
+			logger.Error(err, "unable to scan path")
```

Comment:
The error message could better describe what failed. Maybe something like `unable to get file info`